### PR TITLE
[2021.2] [ShaderGraph] [bugfix 1301915] Node Include files are now tracked explicitly BY GUID

### PR DESCRIPTION
**Pull request by cdxntchou**
### Purpose of this PR
Fix for https://fogbugz.unity3d.com/f/cases/1301915/

Include files are now tracked by GUID (so they can find a file even after it moves), the includes are propagated out of subgraphs, and merged uniquely by the generator.

Some care was taken to keep the HDRP "replace placeholder" functionality working properly.

---
### Testing status
- [x] Verified that moving or renaming CFN-include files from a subgraph now works correctly -- the subgraph does NOT reimport, but the graphs that use it DO reimport, and the correct path is used
- [x] Verified that serialized subgraph assets contains include information from custom function nodes
- [x] Verified that serialized subgraph assets contains include information from parallax mapping nodes
- [x] Verified that subgraph nodes pull in includes from subgraph assets
- [x] Verified that multiple-includes of the same file get de-duplicated
- [x] Verified that custom function nodes using includes still work correctly
- [x] Verified that parallax mapping nodes still work correctly
image
- [x] verified that HDRP "replace placeholder" include continues to work properly
- [x] tested VFX integration path, when using CFN-include files
- [x] also tested moving directories of those files used by VFX
image

Added new Editor test for moving directories of ShaderGraph/SubGraphs containing CFN nodes using include files, similar to the repro case.
- [x] verified that the test failed before this PR
- [x] verified that it now works, after this PR

ShaderGraph PR job: 🟢 
https://yamato.cds.internal.unity3d.com/jobs/902-Graphics/tree/sg%252Ffix%252F1301915/.yamato%252Fall-shadergraph.yml%2523PR_ShaderGraph_trunk/5141106/job/pipeline

ShaderGraph Stereo PR job: 🟢 
https://yamato.cds.internal.unity3d.com/jobs/902-Graphics/tree/sg%252Ffix%252F1301915/.yamato%252Fall-shadergraph_stereo.yml%2523PR_ShaderGraph_Stereo_trunk/5141105/job/pipeline

HDRP PR job:  failures match master failures
https://yamato.cds.internal.unity3d.com/jobs/902-Graphics/tree/sg%252Ffix%252F1301915/.yamato%252Fall-hdrp.yml%2523PR_HDRP_trunk/5141136/job/pipeline

Master equivalent job:
https://yamato.cds.internal.unity3d.com/jobs/902-Graphics/tree/master/.yamato%252Fall-hdrp.yml%2523PR_HDRP_trunk/5068229/job/pipeline

Failures:
HDRP on Linux_Vulkan_playmode_cache_mono_Linear on version trunk -- same 4 tests failed same way
HDRP on OSX_Metal_playmode_cache_mono_Linear on version trunk -- master failed 8 tests, this branch failed 6, all matching master failures
  master additionally failed: HDRP_GraphicTestRunner.2106_GI_EmissionSG, HDRP_GraphicTestRunner.2108_GI_IESCookie_B
seems unrelated to my changes, I assume non-determinism
HDRP on Win_Vulkan_playmode_cache_mono_Linear on version trunk -- same 4 tests failed same way
HDRP on Win_DX12_playmode_cache_mono_Linear on version trunk  -- same 5 tests failed same way
    slight difference on error amount for one of the same tests that failed in master OSX_Metal above -- assume non-determinism
HDRP on Win_DX11_playmode_XR_cache_mono_Linear on version trunk -- both had script compile failures
HDRP on Win_DX11_playmode_cache_mono_Linear on version trunk -- same 4 tests failed same way

HDRP_DXR PR Job: 🟢 
https://yamato.cds.internal.unity3d.com/jobs/902-Graphics/tree/sg%252Ffix%252F1301915/.yamato%252Fall-hdrp_dxr.yml%2523PR_HDRP_DXR_trunk/5141127/job/pipeline

HDRP_Hybrid PR Job: 🟢 
https://yamato.cds.internal.unity3d.com/jobs/902-Graphics/tree/sg%252Ffix%252F1301915/.yamato%252Fall-hdrp_hybrid.yml%2523PR_HDRP_Hybrid_trunk/5141129/job/pipeline

HDRP VFX PR job: 🟢 
https://yamato.cds.internal.unity3d.com/jobs/902-Graphics/tree/sg%252Ffix%252F1301915/.yamato%252Fall-vfx_hdrp.yml%2523PR_VFX_HDRP_trunk/5141135/job/pipeline

Universal PR job:  failures match master failures
https://yamato.cds.internal.unity3d.com/jobs/902-Graphics/tree/sg%252Ffix%252F1301915/.yamato%252Fall-universal.yml%2523PR_Universal_trunk/5174976/job/pipeline

Master equivalent job:
https://yamato.cds.internal.unity3d.com/jobs/902-Graphics/tree/master/.yamato%252Fall-universal.yml%2523PR_Universal_trunk/5068179/job/pipeline

failures:
Universal on iPhone_Metal_Standalone_cache_il2cpp_Linear on version trunk -- same 1 test failed same way
Universal on Android_Vulkan_Standalone_cache_il2cpp_Linear on version trunk -- did not complete, re-running
Universal on Android_OpenGLES3_Standalone_cache_il2cpp_Linear on version trunk -- same 3 tests failed same way
Universal on Win_DX11_playmode_XR_cache_mono_Linear on version trunk -- both had script compile errors

---
### Comments to reviewers
Notes for the reviewers you have assigned.

### CODE REVIEW

File: com.unity.render-pipelines.high-definition/Editor/Material/ShaderGraph/HDSubTarget.cs [hunk 1]
```diff
@@ -139,22 +139,23 @@ protected SubShaderDescriptor PostProcessSubShader(SubShaderDescriptor subShader
...
-                    if (include.descriptor.value == CoreIncludes.kPostDecalsPlaceholder)
-                        include.descriptor.value = postDecalsInclude;
-                    if (include.descriptor.value == CoreIncludes.kRaytracingPlaceholder)
-                        include.descriptor.value = raytracingInclude;
-                    if (include.descriptor.value == CoreIncludes.kPathtracingPlaceholder)
-                        include.descriptor.value = pathtracingInclude;
-
-                    if (!String.IsNullOrEmpty(include.descriptor.value))
-                        finalIncludes.Add(include.descriptor.value, include.descriptor.location, include.fieldConditions);
+                    var path = include.path;
```
Comment (cdxntchou) [inline com.unity.render-pipelines.high-definition/Editor/Material/ShaderGraph/HDSubTarget.cs:146]: This seems like an error in that it used to be modifying the values stored in the passDescriptor (which may be static global)..   now it should not modify the existing structure...

File: com.unity.shadergraph/Editor/Data/Nodes/Utility/CustomFunctionNode.cs [hunk 2]
```diff
@@ -422,9 +411,9 @@ public override void ValidateNode()
             base.ValidateNode();
         }
 
-        public bool Reload(HashSet<string> changedFileDependencies)
+        public bool Reload(HashSet<string> changedFileDependencyGUIDs)
```
Comment (cdxntchou) [inline com.unity.shadergraph/Editor/Data/Nodes/Utility/CustomFunctionNode.cs:414]: this is just changing the name of the variable to make it clear these are GUIDs not paths

File: com.unity.shadergraph/Editor/Data/Nodes/Utility/SubGraphNode.cs [hunk 3]
```diff
@@ -564,6 +564,8 @@ public virtual void GenerateNodeFunction(FunctionRegistry registry, GenerationMo
             if (asset == null || hasError)
                 return;
 
+            registry.RequiresIncludes(asset.includes);
```
Comment (cdxntchou) [inline com.unity.shadergraph/Editor/Data/Nodes/Utility/SubGraphNode.cs:567]: This forwards the includes from the subgraph asset to any graph using them..

File: com.unity.shadergraph/Editor/Data/Util/FunctionRegistry.cs [hunk 4]
```diff
@@ -15,10 +15,12 @@ class FunctionRegistry
         Dictionary<string, FunctionSource> m_Sources = new Dictionary<string, FunctionSource>();
         bool m_Validate = false;
         ShaderStringBuilder m_Builder;
+        IncludeCollection m_Includes;
```
Comment (cdxntchou) [inline com.unity.shadergraph/Editor/Data/Util/FunctionRegistry.cs:18]: Easiest was to make the FunctionRegistry effectively a "FunctionAndNodeInclude" registry...

File: com.unity.shadergraph/Editor/Generation/Collections/DefineCollection.cs [hunk 5]
```diff
@@ -6,7 +6,7 @@ namespace UnityEditor.ShaderGraph
     [GenerationAPI]
     internal class DefineCollection : IEnumerable<DefineCollection.Item>
     {
-        public class Item : IConditional, IShaderString
```
Comment (cdxntchou) [inline com.unity.shadergraph/Editor/Generation/Collections/DefineCollection.cs:9]: IShaderString interface is not used anywhere...  I removed it.

File: com.unity.shadergraph/Editor/Generation/Collections/IncludeCollection.cs [hunk 6]
```diff
@@ -1,61 +1,85 @@
+using System;
 using System.Collections;
 using System.Collections.Generic;
+using UnityEngine;
 
 namespace UnityEditor.ShaderGraph
 {
     [GenerationAPI]
-    internal class IncludeCollection : IEnumerable<IncludeCollection.Item>
+    [Serializable]
```
Comment (cdxntchou) [inline com.unity.shadergraph/Editor/Generation/Collections/IncludeCollection.cs:9]: The changes here are to make the IncludeCollection serializable, so we can serialize it as part of the subgraph asset.

Comment (cdxntchou) [reply com.unity.shadergraph/Editor/Generation/Collections/IncludeCollection.cs:9]: Note that I removed "Item" and just combined it with the IncludeDescriptor.  No one else was using IncludeDescriptor, so there didn't seem to be any reason to have separate classes for the two.

File: com.unity.shadergraph/Editor/Generation/Enumerations/IncludeLocation.cs [hunk 7]
```diff
@@ -1,6 +1,9 @@
+using System;
+
 namespace UnityEditor.ShaderGraph
 {
     [GenerationAPI]
+    [Serializable]
```
Comment (cdxntchou) [inline com.unity.shadergraph/Editor/Generation/Enumerations/IncludeLocation.cs:6]: Not strictly necessary, enums are serializable by default, but serves as a good reminder that this is a serialized enumeration, if anyone is looking to change it in the future...

File: com.unity.shadergraph/Editor/Generation/Processors/Generator.cs [hunk 8]
```diff
@@ -413,36 +414,6 @@ void GenerateShaderPass(int targetIndex, PassDescriptor pass, ActiveFields activ
                 spliceCommands.Add("PassPragmas", command);
             }
```
Comment (cdxntchou) [inline com.unity.shadergraph/Editor/Generation/Processors/Generator.cs:416]: These move further down, so we can combine the Target/pass includes with the node includes before generating them.

**Review by elizabeth-legros — APPROVED** [review]

**Review by LandonTownsendUnity — APPROVED** [review]
Did just a bit of extra testing (setting up a use case scenario and making sure the bug is fixed). Other than that, developer testing is sufficient.

**Pull request merged**